packages/graphql/src/schema/validation/custom-rules/features/valid-relationship-declaration.ts

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright (c) "Neo4j"
+ * Neo4j Sweden AB [http://neo4j.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type {
+    ASTVisitor,
+    DirectiveNode,
+    InterfaceTypeDefinitionNode,
+    ObjectTypeDefinitionNode,
+    ObjectTypeExtensionNode,
+} from "graphql";
+import { visit, Kind } from "graphql";
+import type { ASTValidationContext, SDLValidationContext } from "graphql/validation/ValidationContext";
+import type { ObjectOrInterfaceDefinitionNode } from "../../enrichers/directive/utils";
+import { assertValid, createGraphQLError, DocumentValidationError } from "../utils/document-validation-error";
+import {
+    getInheritedTypeNames,
+    hydrateInterfaceWithImplementedTypesMap,
+} from "../utils/interface-to-implementing-types";
+import type { ObjectOrInterfaceWithExtensions } from "../utils/path-parser";
+import { getPathToNode } from "../utils/path-parser";
+
+export function ValidRelationshipDeclaration(context: SDLValidationContext): ASTVisitor {
```

Comment:
As mentioned, I would suggest trying to reduce the size of this function by delegating some of this logic to external functions (for example, creating the nested Visitor)